File: Platform/infra/security/interfaces/dotnet/AuthInterfaces.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Coyote.Infra.Security.Auth;

/// <summary>
/// Multi-standard authentication modes supported by the platform
/// </summary>
public enum AuthMode
{    /// <summary>
    /// Standard OAuth2 client credentials flow (RFC 6749)
    /// </summary>
    ClientCredentials,
    
    /// <summary>
    /// Client credentials with mutual TLS authentication (RFC 8705)
    /// </summary>
    ClientCredentialsMtls,
    
    /// <summary>
    /// JWT Bearer assertion flow (RFC 7523)
    /// </summary>
    JwtBearer,
    
    /// <summary>
    /// Authorization code flow (RFC 6749)
    /// </summary>
    AuthorizationCode,
    
    /// <summary>
    /// Authorization code flow with PKCE (RFC 7636)
    /// </summary>
    AuthorizationCodePkce
}

/// <summary>
/// Multi-standard authentication client configuration supporting OAuth2 (RFC 6749),
/// JWT Bearer (RFC 7523), and mTLS (RFC 8705) authentication methods.
/// </summary>
public class AuthClientConfig
{
    /// <summary>
    /// Authentication mode to use
    /// </summary>
    public AuthMode AuthMode { get; set; } = AuthMode.ClientCredentials;    /// <summary>
    /// Authentication server base URL
    /// </summary>
    public string ServerUrl { get; set; } = string.Empty;

    /// <summary>
    /// Authentication client ID
    /// </summary>
    public string ClientId { get; set; } = string.Empty;

    /// <summary>
    /// Authentication client secret (for confidential clients)
    /// </summary>
    public string? ClientSecret { get; set; }

    /// <summary>
    /// Default scopes to request
    /// </summary>
    public List<string> DefaultScopes { get; set; } = new();

    /// <summary>
    /// Client certificate path for mTLS authentication
    /// </summary>
    public string? ClientCertPath { get; set; }

    /// <summary>
    /// Client certificate key path for mTLS authentication
    /// </summary>
    public string? ClientKeyPath { get; set; }

    /// <summary>
    /// CA certificate path for mTLS authentication
    /// </summary>
    public string? CaCertPath { get; set; }

    /// <summary>
    /// JWT signing key path for JWT Bearer flow
    /// </summary>
    public string? JwtSigningKeyPath { get; set; }

    /// <summary>
    /// JWT algorithm for JWT Bearer flow
    /// </summary>
    public string JwtAlgorithm { get; set; } = "RS256";

    /// <summary>
    /// JWT issuer for JWT Bearer flow
    /// </summary>
    public string? JwtIssuer { get; set; }

    /// <summary>
    /// JWT audience for JWT Bearer flow
    /// </summary>
    public string? JwtAudience { get; set; }

    /// <summary>
    /// Redirect URI for authorization code flows
    /// </summary>
    public string? RedirectUri { get; set; }

    /// <summary>
    /// Use PKCE for authorization code flow
    /// </summary>
    public bool UsePkce { get; set; } = true;

    /// <summary>
    /// Token refresh buffer in seconds (refresh before expiry)
    /// </summary>
    public int RefreshBufferSeconds { get; set; } = 300; // 5 minutes

    /// <summary>
    /// Enable automatic token refresh
    /// </summary>
    public bool AutoRefresh { get; set; } = true;

    /// <summary>
    /// Maximum retry attempts for token operations
    /// </summary>
    public int MaxRetryAttempts { get; set; } = 3;

    /// <summary>
    /// Retry delay in milliseconds
    /// </summary>
    public int RetryDelayMs { get; set; } = 1000;

    /// <summary>
    /// HTTP timeout in milliseconds
    /// </summary>
    public int TimeoutMs { get; set; } = 30000;

    /// <summary>
    /// Verify SSL certificates
    /// </summary>
    public bool VerifySsl { get; set; } = true;

    // Helper methods for authentication mode checking
    
    /// <summary>
    /// Check if using client credentials mode
    /// </summary>
    public bool IsClientCredentialsMode() => AuthMode == AuthMode.ClientCredentials;

    /// <summary>
    /// Check if using mTLS mode
    /// </summary>
    public bool IsMtlsMode() => AuthMode == AuthMode.ClientCredentialsMtls;

    /// <summary>
    /// Check if using JWT Bearer mode
    /// </summary>
    public bool IsJwtBearerMode() => AuthMode == AuthMode.JwtBearer;

    /// <summary>
    /// Check if using any authorization code mode
    /// </summary>
    public bool IsAuthorizationCodeMode() => 
        AuthMode == AuthMode.AuthorizationCode || AuthMode == AuthMode.AuthorizationCodePkce;

    /// <summary>
    /// Check if certificates are required for this mode
    /// </summary>
    public bool RequiresCertificates() => IsMtlsMode();

    /// <summary>
    /// Check if client secret is required for this mode
    /// </summary>
    public bool RequiresClientSecret() => IsClientCredentialsMode() || IsMtlsMode();

    /// <summary>
    /// Check if JWT key is required for this mode
    /// </summary>
    public bool RequiresJwtKey() => IsJwtBearerMode();

    /// <summary>
    /// Check if redirect URI is required for this mode
    /// </summary>
    public bool RequiresRedirectUri() => IsAuthorizationCodeMode();

    /// <summary>
    /// Validate configuration for the selected authentication mode
    /// </summary>
    public bool IsValid()
    {
        if (string.IsNullOrEmpty(ClientId) || string.IsNullOrEmpty(ServerUrl))
        {
            return false;
        }

        return AuthMode switch
        {
            AuthMode.ClientCredentials => !string.IsNullOrEmpty(ClientSecret),
            AuthMode.ClientCredentialsMtls => !string.IsNullOrEmpty(ClientSecret) && 
                                            !string.IsNullOrEmpty(ClientCertPath) && 
                                            !string.IsNullOrEmpty(ClientKeyPath),
            AuthMode.JwtBearer => !string.IsNullOrEmpty(JwtSigningKeyPath),
            AuthMode.AuthorizationCode or AuthMode.AuthorizationCodePkce => !string.IsNullOrEmpty(RedirectUri),
            _ => false
        };
    }
}

/// <summary>
/// Authentication token storage interface
/// </summary>
public interface IAuthTokenStorage
{
    /// <summary>
    /// Store a token for a client
    /// </summary>
    Task StoreTokenAsync(string clientId, AuthToken token);

    /// <summary>
    /// Retrieve a token for a client
    /// </summary>
    AuthToken? GetToken(string clientId);

    /// <summary>
    /// Clear stored token for a client
    /// </summary>
    void ClearToken(string clientId);

    /// <summary>
    /// Clear all stored tokens
    /// </summary>
    void ClearAllTokens();
}

/// <summary>
/// Authentication logger interface
/// </summary>
public interface IAuthLogger
{
    /// <summary>
    /// Log information message
    /// </summary>
    void LogInfo(string message);

    /// <summary>
    /// Log error message
    /// </summary>
    void LogError(string message);

    /// <summary>
    /// Log debug message
    /// </summary>
    void LogDebug(string message);
}

/// <summary>
/// In-memory token storage implementation
/// </summary>
public class InMemoryTokenStorage : IAuthTokenStorage
{    private readonly System.Collections.Concurrent.ConcurrentDictionary<string, AuthToken> _tokens = new();

    /// <summary>
    /// Stores an authentication token for the specified client
    /// </summary>
    public Task StoreTokenAsync(string clientId, AuthToken token)
    {
        _tokens[clientId] = token;
        return Task.CompletedTask;
    }

    /// <summary>
    /// Retrieves the stored token for the specified client
    /// </summary>
    public AuthToken? GetToken(string clientId)
    {
        return _tokens.TryGetValue(clientId, out var token) ? token : null;
    }

    /// <summary>
    /// Clears the stored token for the specified client
    /// </summary>
    public void ClearToken(string clientId)
    {
        _tokens.TryRemove(clientId, out _);
    }

    /// <summary>
    /// Clears all stored tokens
    /// </summary>
    public void ClearAllTokens()
    {
        _tokens.Clear();
    }
}

/// <summary>
/// Console logger implementation
/// </summary>
public class ConsoleAuthLogger : IAuthLogger
{    private readonly string _prefix;

    /// <summary>
    /// Initializes a new instance of the ConsoleAuthLogger
    /// </summary>
    public ConsoleAuthLogger(string? prefix = null)
    {
        _prefix = prefix ?? "Auth";
    }

    /// <summary>
    /// Logs an informational message to the console
    /// </summary>
    public void LogInfo(string message)
    {
        Console.WriteLine($"[{DateTime.UtcNow:yyyy-MM-dd HH:mm:ss}] [{_prefix}] INFO: {message}");
    }

    /// <summary>
    /// Logs an error message to the console
    /// </summary>
    public void LogError(string message)
    {        Console.WriteLine($"[{DateTime.UtcNow:yyyy-MM-dd HH:mm:ss}] [{_prefix}] ERROR: {message}");
    }

    /// <summary>
    /// Logs a debug message to the console
    /// </summary>
    public void LogDebug(string message)
    {
        Console.WriteLine($"[{DateTime.UtcNow:yyyy-MM-dd HH:mm:ss}] [{_prefix}] DEBUG: {message}");
    }
}

/// <summary>
/// Null logger implementation (no logging)
/// </summary>
public class NullAuthLogger : IAuthLogger
{
    /// <summary>
    /// Logs an informational message (no-op)
    /// </summary>
    public void LogInfo(string message) { }
    
    /// <summary>
    /// Logs an error message (no-op)
    /// </summary>
    public void LogError(string message) { }
    
    /// <summary>
    /// Logs a debug message (no-op)
    /// </summary>
    public void LogDebug(string message) { }
}

/// <summary>
/// Authentication token information
/// </summary>
public class AuthToken
{
    /// <summary>
    /// Access token
    /// </summary>
    public string AccessToken { get; set; } = string.Empty;

    /// <summary>
    /// Token type (usually "Bearer")
    /// </summary>
    public string TokenType { get; set; } = "Bearer";

    /// <summary>
    /// Token expiration time
    /// </summary>
    public DateTime ExpiresAt { get; set; }

    /// <summary>
    /// Refresh token (if available)
    /// </summary>
    public string? RefreshToken { get; set; }

    /// <summary>
    /// Token scopes
    /// </summary>
    public List<string> Scopes { get; set; } = new();

    /// <summary>
    /// Check if token is expired
    /// </summary>
    public bool IsExpired => DateTime.UtcNow >= ExpiresAt;

    /// <summary>
    /// Check if token needs refresh (within buffer time)
    /// </summary>
    public bool NeedsRefresh(int bufferSeconds = 300) => DateTime.UtcNow.AddSeconds(bufferSeconds) >= ExpiresAt;

    /// <summary>
    /// Get authorization header value
    /// </summary>
    public string GetAuthorizationHeader() => $"{TokenType} {AccessToken}";
}

/// <summary>
/// Authentication result
/// </summary>
public class AuthResult
{
    /// <summary>
    /// Whether authentication was successful
    /// </summary>
    public bool IsSuccess { get; set; }

    /// <summary>
    /// Authentication token (if successful)
    /// </summary>
    public AuthToken? Token { get; set; }

    /// <summary>
    /// Error code (if failed)
    /// </summary>
    public string? ErrorCode { get; set; }

    /// <summary>
    /// Error description (if failed)
    /// </summary>
    public string? ErrorDescription { get; set; }

    /// <summary>
    /// Additional error details
    /// </summary>
    public string? ErrorDetails { get; set; }

    /// <summary>
    /// Create success result
    /// </summary>
    public static AuthResult Success(AuthToken token) => new()
    {
        IsSuccess = true,
        Token = token
    };

    /// <summary>
    /// Create error result
    /// </summary>
    public static AuthResult Error(string errorCode, string? errorDescription = null, string? errorDetails = null) => new()
    {
        IsSuccess = false,
        ErrorCode = errorCode,
        ErrorDescription = errorDescription,
        ErrorDetails = errorDetails
    };
}

/// <summary>
/// Authentication server information
/// </summary>
public class AuthServerInfo
{
    /// <summary>
    /// Authorization endpoint URL
    /// </summary>
    public string AuthorizationEndpoint { get; set; } = string.Empty;

    /// <summary>
    /// Token endpoint URL
    /// </summary>
    public string TokenEndpoint { get; set; } = string.Empty;

    /// <summary>
    /// Token introspection endpoint URL
    /// </summary>
    public string? IntrospectionEndpoint { get; set; }

    /// <summary>
    /// Token revocation endpoint URL
    /// </summary>
    public string? RevocationEndpoint { get; set; }

    /// <summary>
    /// Supported grant types
    /// </summary>
    public List<string> GrantTypesSupported { get; set; } = new();

    /// <summary>
    /// Supported scopes
    /// </summary>
    public List<string> ScopesSupported { get; set; } = new();
}

File: Platform/infra/security/src/ts/interfaces/auth-interfaces.ts
/**
 * Multi-Standard Authentication Interfaces and Types for TypeScript/JavaScript
 * 
 * This module contains all the interface definitions and types for multi-standard authentication
 * in the CoyoteSense platform, supporting OAuth2 (RFC 6749), JWT Bearer (RFC 7523), 
 * and mTLS (RFC 8705) authentication methods.
 */

// Authentication Modes

/**
 * Authentication modes supported by the platform
 */
export enum AuthMode {
  /** Standard OAuth2 client credentials flow (RFC 6749) */
  ClientCredentials = 'client_credentials',
  
  /** Client credentials with mutual TLS authentication (RFC 8705) */
  ClientCredentialsMtls = 'client_credentials_mtls',
  
  /** JWT Bearer assertion flow (RFC 7523) */
  JwtBearer = 'jwt_bearer',
  
  /** Authorization code flow (RFC 6749) */
  AuthorizationCode = 'authorization_code',
  
  /** Authorization code flow with PKCE (RFC 7636) */
  AuthorizationCodePkce = 'authorization_code_pkce'
}

// Authentication Configuration and Models

export interface AuthClientConfig {
  /** Authentication mode to use */
  authMode?: AuthMode;
  /** Authentication server base URL */
  serverUrl: string;
  /** Client ID for authentication */
  clientId: string;
  /** Client secret for authentication (optional for public clients) */
  clientSecret?: string;
  /** Default scopes to request */
  defaultScopes?: string[];
  
  // mTLS settings
  /** Client certificate path for mTLS authentication */
  clientCertPath?: string;
  /** Client certificate key path for mTLS authentication */
  clientKeyPath?: string;
  /** CA certificate path for mTLS authentication */
  caCertPath?: string;
  
  // JWT Bearer settings
  /** JWT signing key path for JWT Bearer flow */
  jwtSigningKeyPath?: string;
  /** JWT algorithm for JWT Bearer flow */
  jwtAlgorithm?: string;
  /** JWT issuer for JWT Bearer flow */
  jwtIssuer?: string;
  /** JWT audience for JWT Bearer flow */
  jwtAudience?: string;
  
  // Authorization Code settings
  /** Redirect URI for authorization code flows */
  redirectUri?: string;
  /** Use PKCE for authorization code flow */
  usePkce?: boolean;
  
  // Token management
  /** Token refresh buffer in seconds (refresh before expiry) */
  refreshBufferSeconds?: number;
  /** Enable automatic token refresh */
  autoRefresh?: boolean;
  /** Maximum retry attempts for token operations */
  maxRetryAttempts?: number;
  /** Retry delay in milliseconds */
  retryDelayMs?: number;
  
  // HTTP settings
  /** Default timeout for authentication requests in milliseconds */
  timeoutMs?: number;
  /** Verify SSL certificates */
  verifySsl?: boolean;
  /** Custom headers to include in authentication requests */
  customHeaders?: Record<string, string>;
  /** Enable automatic token refresh */
  enableAutoRefresh?: boolean;
  /** Token refresh margin in seconds (refresh when token expires in this time) */
  refreshMarginSeconds?: number;
  /** Background refresh check interval in milliseconds */
  refreshCheckIntervalMs?: number;
}

/**
 * Authentication client configuration helper functions
 */
export class AuthConfigHelper {
  /** Check if using client credentials mode */
  static isClientCredentialsMode(config: AuthClientConfig): boolean {
    return config.authMode === AuthMode.ClientCredentials;
  }

  /** Check if using mTLS mode */
  static isMtlsMode(config: AuthClientConfig): boolean {
    return config.authMode === AuthMode.ClientCredentialsMtls;
  }

  /** Check if using JWT Bearer mode */
  static isJwtBearerMode(config: AuthClientConfig): boolean {
    return config.authMode === AuthMode.JwtBearer;
  }

  /** Check if using any authorization code mode */
  static isAuthorizationCodeMode(config: AuthClientConfig): boolean {
    return config.authMode === AuthMode.AuthorizationCode || 
           config.authMode === AuthMode.AuthorizationCodePkce;
  }

  /** Check if certificates are required for this mode */
  static requiresCertificates(config: AuthClientConfig): boolean {
    return this.isMtlsMode(config);
  }

  /** Check if client secret is required for this mode */
  static requiresClientSecret(config: AuthClientConfig): boolean {
    return this.isClientCredentialsMode(config) || this.isMtlsMode(config);
  }

  /** Check if JWT key is required for this mode */
  static requiresJwtKey(config: AuthClientConfig): boolean {
    return this.isJwtBearerMode(config);
  }

  /** Check if redirect URI is required for this mode */
  static requiresRedirectUri(config: AuthClientConfig): boolean {
    return this.isAuthorizationCodeMode(config);
  }

  /** Validate configuration for the selected authentication mode */
  static isValid(config: AuthClientConfig): boolean {
    if (!config.clientId || !config.serverUrl) {
      return false;
    }

    switch (config.authMode) {
      case AuthMode.ClientCredentials:
        return !!config.clientSecret;
      
      case AuthMode.ClientCredentialsMtls:
        return !!config.clientSecret && !!config.clientCertPath && !!config.clientKeyPath;
      
      case AuthMode.JwtBearer:
        return !!config.jwtSigningKeyPath;
      
      case AuthMode.AuthorizationCode:
      case AuthMode.AuthorizationCodePkce:
        return !!config.redirectUri;
      
      default:
        return false;
    }
  }
}

export interface AuthToken {
  /** Access token value */
  accessToken: string;
  /** Token type (typically 'Bearer') */
  tokenType: string;
  /** Token expiration time as Unix timestamp */
  expiresAt: number;
  /** Refresh token (if available) */
  refreshToken?: string;
  /** Token scopes */
  scopes?: string[];
  /** ID token (for OpenID Connect) */
  idToken?: string;
}

export interface AuthResult {
  /** Whether the authentication was successful */
  success: boolean;
  /** Access token (if successful) */
  token?: AuthToken;
  /** Error code (if failed) */
  error?: string;
  /** Error description (if failed) */
  errorDescription?: string;
  /** Server info (if available) */
  serverInfo?: AuthServerInfo;
}

export interface AuthServerInfo {
  /** Authentication server issuer */
  issuer?: string;
  /** Authorization endpoint URL */
  authorizationEndpoint?: string;
  /** Token endpoint URL */
  tokenEndpoint?: string;
  /** Token introspection endpoint URL */
  introspectionEndpoint?: string;
  /** Token revocation endpoint URL */
  revocationEndpoint?: string;
  /** Supported grant types */
  grantTypesSupported?: string[];
  /** Supported response types */
  responseTypesSupported?: string[];
  /** Supported scopes */
  scopesSupported?: string[];
}

export interface AuthPKCEData {
  /** Code verifier for PKCE */
  codeVerifier: string;
  /** Code challenge for PKCE */
  codeChallenge: string;
  /** Code challenge method (S256 or plain) */
  codeChallengeMethod: string;
  /** Authorization state parameter */
  state: string;
}

// Authentication Request/Response Models (matching server API)

export interface AuthTokenRequest {
  grant_type: string;
  client_id?: string;
  client_secret?: string;
  scope?: string;
  assertion?: string;
  code?: string;
  redirect_uri?: string;
  refresh_token?: string;
  code_verifier?: string;
}

export interface AuthTokenResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
  refresh_token?: string;
  scope?: string;
  id_token?: string;
}

export interface AuthErrorResponse {
  error: string;
  error_description?: string;
  error_uri?: string;
}

export interface AuthIntrospectRequest {
  token: string;
  token_type_hint?: string;
}

export interface AuthIntrospectResponse {
  active: boolean;
  scope?: string;
  client_id?: string;
  username?: string;
  token_type?: string;
  exp?: number;
  iat?: number;
  sub?: string;
  aud?: string;
  iss?: string;
  jti?: string;
}

// Abstract Interfaces

export abstract class AuthTokenStorage {
  /** Store a token with the specified key */
  abstract storeTokenAsync(key: string, token: AuthToken): Promise<void>;
  
  /** Retrieve a token by key */
  abstract retrieveTokenAsync(key: string): Promise<AuthToken | null>;
  
  /** Remove a token by key */
  abstract removeTokenAsync(key: string): Promise<void>;
  
  /** Clear all stored tokens */
  abstract clearAllTokensAsync(): Promise<void>;
}

export abstract class AuthLogger {
  /** Log debug message */
  abstract debug(message: string, ...args: any[]): void;
  
  /** Log info message */
  abstract info(message: string, ...args: any[]): void;
  
  /** Log warning message */
  abstract warn(message: string, ...args: any[]): void;
  
  /** Log error message */
  abstract error(message: string, ...args: any[]): void;
}

// Concrete Implementations

export class MemoryAuthTokenStorage extends AuthTokenStorage {
  private readonly tokens = new Map<string, AuthToken>();

  async storeTokenAsync(key: string, token: AuthToken): Promise<void> {
    this.tokens.set(key, { ...token });
  }

  async retrieveTokenAsync(key: string): Promise<AuthToken | null> {
    const token = this.tokens.get(key);
    return token ? { ...token } : null;
  }

  async removeTokenAsync(key: string): Promise<void> {
    this.tokens.delete(key);
  }

  async clearAllTokensAsync(): Promise<void> {
    this.tokens.clear();
  }
}

export class ConsoleAuthLogger extends AuthLogger {
  constructor(private readonly prefix: string = '[AuthClient]') {
    super();
  }

  debug(message: string, ...args: any[]): void {
    console.debug(`${this.prefix} DEBUG: ${message}`, ...args);
  }

  info(message: string, ...args: any[]): void {
    console.info(`${this.prefix} INFO: ${message}`, ...args);
  }

  warn(message: string, ...args: any[]): void {
    console.warn(`${this.prefix} WARN: ${message}`, ...args);
  }

  error(message: string, ...args: any[]): void {
    console.error(`${this.prefix} ERROR: ${message}`, ...args);
  }
}

export class NullAuthLogger extends AuthLogger {
  debug(message: string, ...args: any[]): void {}
  info(message: string, ...args: any[]): void {}
  warn(message: string, ...args: any[]): void {}
  error(message: string, ...args: any[]): void {}
}

// Main Authentication Client Interface

export interface IAuthClient {
  // Client Credentials Flow
  /**
   * Authenticate using Client Credentials flow (OAuth2 RFC 6749)
   * @param scopes Requested scopes
   * @returns Authentication result
   */
  clientCredentialsAsync(scopes?: string[]): Promise<AuthResult>;

  // JWT Bearer Flow
  /**
   * Authenticate using JWT Bearer assertion (RFC 7523)
   * @param assertion JWT assertion token
   * @param scopes Requested scopes
   * @returns Authentication result
   */
  jwtBearerAsync(assertion: string, scopes?: string[]): Promise<AuthResult>;

  // Authorization Code Flow
  /**
   * Start Authorization Code flow by generating authorization URL (OAuth2 RFC 6749)
   * @param redirectUri Redirect URI after authorization
   * @param scopes Requested scopes
   * @param state Optional state parameter
   * @param usePKCE Whether to use PKCE (RFC 7636)
   * @returns Authorization URL and PKCE data
   */
  startAuthorizationCodeFlow(
    redirectUri: string, 
    scopes?: string[], 
    state?: string, 
    usePKCE?: boolean
  ): { authorizationUrl: string; pkceData?: AuthPKCEData };

  /**
   * Complete Authorization Code flow by exchanging code for token
   * @param code Authorization code received from callback
   * @param redirectUri Redirect URI used in authorization request
   * @param pkceData PKCE data from startAuthorizationCodeFlow (if used)
   * @returns Authentication result
   */
  completeAuthorizationCodeFlow(
    code: string, 
    redirectUri: string, 
    pkceData?: AuthPKCEData
  ): Promise<AuthResult>;

  // Token Refresh
  /**
   * Refresh an access token using refresh token
   * @param refreshToken Refresh token
   * @param scopes Optional scopes to request
   * @returns Authentication result
   */
  refreshTokenAsync(refreshToken: string, scopes?: string[]): Promise<AuthResult>;

  // Token Management
  /**
   * Get stored token for client
   * @returns Stored authentication token or null
   */
  getStoredTokenAsync(): Promise<AuthToken | null>;

  /**
   * Store token for client
   * @param token Token to store
   */
  storeTokenAsync(token: AuthToken): Promise<void>;

  /**
   * Remove stored token for client
   */
  removeStoredTokenAsync(): Promise<void>;

  /**
   * Check if stored token is valid and not expired
   * @param marginSeconds Expiration margin in seconds
   * @returns True if token is valid
   */
  hasValidTokenAsync(marginSeconds?: number): Promise<boolean>;

  // Token Introspection
  /**
   * Introspect a token to check its validity and metadata
   * @param token Token to introspect
   * @param tokenTypeHint Optional hint about token type
   * @returns Token introspection result
   */
  introspectTokenAsync(token: string, tokenTypeHint?: string): Promise<AuthIntrospectResponse>;

  // Token Revocation
  /**
   * Revoke a token
   * @param token Token to revoke
   * @param tokenTypeHint Optional hint about token type
   * @returns True if revocation was successful
   */
  revokeTokenAsync(token: string, tokenTypeHint?: string): Promise<boolean>;

  // Server Discovery
  /**
   * Discover authentication server capabilities
   * @returns Server information
   */
  discoverServerAsync(): Promise<AuthServerInfo>;

  // Auto-refresh Management
  /**
   * Start automatic token refresh
   */
  startAutoRefresh(): void;

  /**
   * Stop automatic token refresh
   */
  stopAutoRefresh(): void;

  /**
   * Check if auto-refresh is enabled
   * @returns True if auto-refresh is running
   */
  isAutoRefreshRunning(): boolean;

  // Client Management
  /**
   * Get client configuration
   * @returns Client configuration
   */
  getConfig(): AuthClientConfig;

  /**
   * Dispose of client resources
   */
  dispose(): void;
}

// Legacy OAuth2 interfaces for backward compatibility (deprecated)
// These will be removed in a future version - use Auth* interfaces instead

/** @deprecated Use AuthMode instead */
export const OAuth2AuthMode = AuthMode;

/** @deprecated Use AuthClientConfig instead */
export type OAuth2ClientConfig = AuthClientConfig;

/** @deprecated Use AuthToken instead */
export type OAuth2Token = AuthToken;

/** @deprecated Use AuthResult instead */
export type OAuth2AuthResult = AuthResult;

/** @deprecated Use AuthServerInfo instead */
export type OAuth2ServerInfo = AuthServerInfo;

/** @deprecated Use AuthPKCEData instead */
export type OAuth2PKCEData = AuthPKCEData;

/** @deprecated Use IAuthClient instead */
export type IOAuth2AuthClient = IAuthClient;

/** @deprecated Use AuthConfigHelper instead */
export const OAuth2ConfigHelper = AuthConfigHelper;
